ZJS sensor: initial OCF servers based on Zephyr.js

ttzeng opened the pull request:
Initial set of ZJS based OCF servers for replicating the SmartHome demo servers on Arduino 101

gvancuts commented on ocf-servers/zjs-servers/README.md:
```diff
@@ -0,0 +1,50 @@
+| temperature.js | n/a | onboard temperature sensor |
+| multiserver.js | D2  | Grove PIR motion sensor |
+|                | D4  | Grove button |
+|                | D7  | Grove buzzer |
```
The table is not rendered correctly. I _think_ it may be because you need spaces between the dashes and the pipe signs.

gvancuts commented on ocf-servers/zjs-servers/led.js:
```diff
@@ -0,0 +1,49 @@
+var ocf = require('ocf');
```
Can you add the license header (similar to the files in the `js-servers` folder)?

gvancuts commented on ocf-servers/zjs-servers/multiserver.js:
```diff
@@ -0,0 +1,142 @@
+/*
```
Can you add the license header (similar to the files in the `js-servers` folder)?

gvancuts commented on ocf-servers/zjs-servers/temperature.js:
```diff
@@ -0,0 +1,67 @@
+var resPathTemperature = '/bmi160/temperature',
```
Can you add the license header (similar to the files in the `js-servers` folder)?

uhofemei replied:
Where is this bmi160 coming from? The grove temperature sensor doesn't report Celsius. It reports an 11 bit raw value that needs to be converted. Same with the ambient light sensor.

ttzeng replied:
This is not the Grove temperature sensor connected to the analog inputs, we could use the **onchip** temperature sensor inside the curie chip. From [this sample code](https://github.com/01org/zephyr.js/blob/master/docs/sensors.md), it sounds the curie chip on Arduino 101 has built-in BMI160 sensors, and ZJS runtime has driver for it to report sensed temperature in Celsius.

uhofemei replied:
Ah that's what the multi sensor (while it's still running) must be using then. It reports 150 F though. Way too hot :)

gvancuts replied:
My understanding is that the value returned by that temp sensor is the temperature of the [BMI160](https://www.bosch-sensortec.com/bst/products/all_products/bmi160) chip itself, not the temperature of the environment. This could explain the values you are seeing.

See also the [original driver from Bosch](https://github.com/BoschSensortec/BMI160_driver/blob/master/bmi160.c#L3172) and the [driver we have in zephyr](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/sensor/bmi160/bmi160.c#L752)

uhofemei replied:
That's correct. I will go with Tonny's PR nonetheless for the staging environment as discussed earlier.

gvancuts commented on ocf-servers/zjs-servers/README.md:
```diff
@@ -0,0 +1,50 @@
+# bluetoothctl
+[bluetooth]# scan on
+[bluetooth]# scan off
+[bluetooth]# devices # You will see the "Zephyr OCF Sensors" with the address
```
Is this the device name the same for all three ZJS examples included in this PR?

ttzeng commented:
@gvancuts Thanks for the review. I've addressed all your inputs, please review the forced push. Thanks.

uhofemei commented:
Hi @ttzeng, could you also work on a ZJS sample code for the ambient light sensor to report lux? Thanks :)

gvancuts commented:
Thanks @ttzeng ! Can someone test this? If positive, this is ready to be merged.

qiuzhong commented:
Tested this PR, here are the results:

* led.js -- onboard
```
$ ./oic-get /res
localhost:8000 /res
[{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/doxm","rt":"oic.r.doxm","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/pstat","rt":"oic.r.pstat","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/d","rt":"oic.wk.d","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"a4506509-0451-4509-5351-6509a2516509","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"a4506509-0451-4509-5351-6509a2516509","links":[{"href":"/oic/d","rt":"oic.d.zjs","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"a4506509-0451-4509-5351-6509a2516509","links":[{"href":"/a/led","rt":"oic.r.switch.bin","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]}]
HTTP: 200
$ ./oic-get /a/led?di=a4506509-0451-4509-5351-6509a2516509
localhost:8000 /a/led?di=a4506509-0451-4509-5351-6509a2516509
{"value":0}
HTTP: 200
```

* temperature.js -- on board
```
./oic-get /res
localhost:8000 /res
[{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/doxm","rt":"oic.r.doxm","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/pstat","rt":"oic.r.pstat","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/d","rt":"oic.wk.d","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"d876720a-3677-420a-4677-720ad677720a","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"d876720a-3677-420a-4677-720ad677720a","links":[{"href":"/oic/d","rt":"oic.d.zjs","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"d876720a-3677-420a-4677-720ad677720a","links":[{"href":"/bmi160/temperature","rt":"oic.r.temperatur","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]}]
HTTP: 200
$ ./oic-get /bmi160/temperature?di=d876720a-3677-420a-4677-720ad677720a
localhost:8000 /bmi160/temperature?di=d876720a-3677-420a-4677-720ad677720a
{"temperature":26.43728,"units":"C"}
HTTP: 200
```

* multisersers.js -- Grove motion/button/buzzer sensors:
```
 ./oic-get /res
localhost:8000 /res
[{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/doxm","rt":"oic.r.doxm","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/sec/pstat","rt":"oic.r.pstat","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/d","rt":"oic.wk.d","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"604ba12a-d289-475d-8a4c-5c44e6c422fe","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.baseline","p":{"bm":1,"secure":false}}]},{"di":"e5c76d09-45c8-4d09-54c8-6d09e3c86d09","links":[{"href":"/oic/p","rt":"oic.wk.p","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"e5c76d09-45c8-4d09-54c8-6d09e3c86d09","links":[{"href":"/oic/d","rt":"oic.d.zjs","if":"oic.if.r","p":{"bm":1,"secure":false}}]},{"di":"e5c76d09-45c8-4d09-54c8-6d09e3c86d09","links":[{"href":"/a/buzzer","rt":"oic.r.buzzer","if":"oic.if.baseline","p":{"bm":3,"secure":false}}]},{"di":"e5c76d09-45c8-4d09-54c8-6d09e3c86d09","links":[{"href":"/a/button","rt":"oic.r.button","if":"oic.if.baseline","p":{"bm":3,"secure":false}}]},{"di":"e5c76d09-45c8-4d09-54c8-6d09e3c86d09","links":[{"href":"/a/pir","rt":"oic.r.sensor.mot","if":"oic.if.baseline","p":{"bm":3,"secure":false}}]}]
HTTP: 200
$ ./oic-get /a/buzzer?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
localhost:8000 /a/buzzer?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
{"value":0}
HTTP: 200
$ ./oic-get /a/pir?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
localhost:8000 /a/pir?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
{"value":1}
HTTP: 200
$ ./oic-get /a/button?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
localhost:8000 /a/button?di=e5c76d09-45c8-4d09-54c8-6d09e3c86d09
{"value":0}
HTTP: 200
```

Here are my concerns:
According to the [OCF spec](https://github.com/01org/SmartHome-Demo/tree/master/ocf-servers/doc) and [suggestion](https://github.com/01org/SmartHome-Demo/issues/79):
* led.js:
  * The type of resource property value should be boolean(true or false), here is numeric(1 or 0)
  * The resource type should be <code>oic.r.switch.binary</code>, here is <code>oic.r.switch.bin</code>
* temperature.js: 
  * The <code>range</code> property is missing.
  * <code>/bmi160/temperature</code> seems a custom href. Generally in all the previous examples, <code>/a/temperature</code> was used. I am not sure if this is a problem.
* multisersers.js:
  * The type of buzzer/button/motion resources property values is still numeric.
  * The resource type should be <code>oic.r.sensor.motion</code>, here is <code>oic.r.sensor.mot</code>
  
Sending requests with POST method to update the sensor states can work well while the value type was modified to accommodate to this PR.

ttzeng commented:
@qiuzhong Thanks for testing and providing valuable feedbacks. Regarding your above comments:
(1) the property value of the led server -- the recently GPIO API changed causes the initial property value didn't set it right. I'll fix this in the next amend PR.
(2) the resource type strings are truncated -- this is [an upstream bug](https://github.com/01org/zephyr.js/issues/1144).
(3) the range property of temperature resource type -- per OIC Resource Type spec, section 6.21, only 'temperature' property is mandatory, the 'range' property can be omitted. If no range is provided, the default is +-MAXINT
(4) the /bmi/temperature resource ID -- per OIC Core Spec 7.3.2.6, resource ID is device and implementation dependent, we can set any ID as long as it's unique across the scope of the host server.

ttzeng commented:
@uhofemei , from the sensor API, it sounds we could leave the illuminance value calculation to the framework code? so I am planning to incorporate [this sample](https://github.com/01org/zephyr.js/blob/master/samples/AmbientLight.js) code to our multiserver script. :)

ttzeng commented:
@qiuzhong I've fixed the issue of assigning numbers to boolean properties, I also added temperature, fan, and light sensor resources to the multiserver script, right now it hosts 6 resources, please take a look.

uhofemei commented:
Thanks Tonny, this is good. I will connect the fan in my demo to the ZJS Arduino as well now.

qiuzhong commented:
@ttzeng , I tried your new update. All the property type issues are fixed now. The sensors work well except the light sensor. The illuminance value I got seems not sensible. 
```
$  ./oic-get /a/illuminance?di=a0bc8f0a-fbbc-4f0a-49bd-8f0a97bd8f0a
localhost:8000 /a/illuminance?di=a0bc8f0a-fbbc-4f0a-49bd-8f0a97bd8f0a
{"illuminance":0.00007453111902153283}
HTTP: 200
``` 

This is not a sensible value in the lab. Generally, it should be sensible if the illuminance is about 20 lux here. I tried other angles for the light sensor but it didn't help.
PS: My light sensor is Grove light sensor v1.1 and it's connected to the A0 port of the Grove Shield.

ttzeng commented:
@qiuzhong , thanks for testing. I guess your Grove base shield was set to 3.3V Vcc?  The illuminance value was calculated and received from the [ZJS runtime](https://github.com/01org/zephyr.js/blob/master/docs/sensors.md), and accordingly to the [sample code](https://github.com/01org/zephyr.js/blob/master/samples/AmbientLight.js), if you set Vcc to 5V, you should get 10000 in daylight mode. I am also wondering whether these lux values make sense, so I'd suggest this should be discussed at [ZJS upstream](https://github.com/01org/zephyr.js).

qiuzhong commented:
@ttzeng , yes, my Grove base shield was set to 3.3V Vcc. If I set it to 5V, the value is about 
<code>{"illuminance":6365.708415668345}</code> in the lab.

gvancuts commented:
I believe this positively concludes the testing of all new features in this PR, is this correct?

@ttzeng , assuming all sensors would work just fine with Vcc set to 5V, can you add a note to that effect in the README.md file?

Thanks!

ttzeng commented:
@gvancuts , I've put a footnote to the README.md to remind the reader that 5V Vcc is required to drive the AREF with proper voltage for the onchip ADC to output correctly.

gvancuts commented:
👍 thanks @ttzeng and @qiuzhong for your help on this, I'll merge this now.
